pkg/controller/expcertificates/issuing/issuing_controller.go

```diff
@@ -0,0 +1,396 @@
+		nowTime := metav1.NewTime(c.clock.Now())
+		crt.Status.LastFailureTime = &nowTime
+
+		var reason, message string
+
+		reason = "FailedToIssue"
+		message = fmt.Sprintf("The certificate request has failed to issue and will be retired: %s",
+			apiutil.CertificateRequestFailedMessage(req))
+
+		crt = crt.DeepCopy()
+		apiutil.SetCertificateCondition(crt, cmapi.CertificateConditionIssuing, cmmeta.ConditionFalse, reason, message)
+
+		if err := c.updateCertificateStatus(ctx, crt); err != nil {
+			return err
+		}
+
+		c.recorder.Event(crt, corev1.EventTypeWarning, "Issuing", message)
+
+		return nil
+
+		// If the CertificateRequest is valid, verify its status and update accordingly.
+	case cmapi.CertificateRequestReasonIssued:
+
+		// Verify the public key of the certificate request matches the public key of the stored status.nextPrivateKeySecretName
+		publicKeyMatches, err := c.certificateRequestPublicKeyMatchesSecret(ctx, req, namespace, nextPrivateKeySecretName)
+		if err != nil {
+			return err
+		}
+
+		// If public key does not match, do nothing (requestmanager will handle this).
+		if !publicKeyMatches {
+			return nil
+		}
+
+		// Verify the CSR options match what is requested in certificate.spec.
+		violations, err := certificates.RequestMatchesSpec(req, crt.Spec)
+		if err != nil {
+			return err
+		}
+
+		// If there are violations in the spec, then the requestmanager will handle this.
+		if len(violations) > 0 {
+			return nil
+		}
+
+		//Encode and issue the key-pair (store it in the Secret)
+		if err := c.issueCertificateAndKeyToSecret(ctx, namespace, crt, req); err != nil {
+			return err
+		}
+
+		//Set status.revision to revision of the CertificateRequest
+		crt.Status.Revision = &nextRevision
+
+		crt = crt.DeepCopy()
+
+		// Remove Issuing status condition
+		apiutil.RemoveCertificateCondition(crt, cmapi.CertificateConditionIssuing)
+
+		//Clear status.lastFailureTime (if set)
+		crt.Status.LastFailureTime = nil
+
+		if err := c.updateCertificateStatus(ctx, crt); err != nil {
+			return err
+		}
+
+		message := "The certificate has been successfully issued"
+		c.recorder.Event(crt, corev1.EventTypeNormal, "Issuing", message)
+
+		return nil
+
+	// CertificateRequest is not in a final state so do nothing.
+	default:
+		return nil
+	}
+}
+
+func (c *controller) updateCertificateStatus(ctx context.Context, crt *cmapi.Certificate) error {
+	updateStatusCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	_, err := c.client.CertmanagerV1alpha2().Certificates(crt.Namespace).UpdateStatus(updateStatusCtx, crt, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// issueCertificateAndKeyToSecret should write the resulting certificate, ca, and private key data to the target secret in the appropriate format type.
+func (c *controller) issueCertificateAndKeyToSecret(ctx context.Context, namespace string, crt *cmapi.Certificate, req *cmapi.CertificateRequest) error {
```

Comment:
Again, I think we can remove this function altogether. Aside from fetching from the lister (which it shouldn't really need to do anyway), the logic here is minimal. We also don't need to have the `if !ok {` check here either if we do this, as we will have already read the key data as well as decoded it by this point.

A simple call to `updateSecretData` from this functions call-site should be sufficient then (and again, more closely aligns with the `Encode and issue the key-pair (store it in the Secret)` item in the design doc, which this function is ultimately implementing right now)